Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,299 @@
+
+When resolving framework reference with a **release** version, prefer release versions. This means that if there's a release version on the machine which satisfies all the requirements it will be chosen, regardless of what pre-release versions are installed. Only if there's no suitable release version, then pre-release versions are also considered (and then they're treated all the same).
+
+Interesting examples:
+* `3.0.0 rollForward = Minor` would not-roll forward and choose `3.0.0` even if `3.0.1-preview` is also available on the machine. This means the automatic roll forward to latest patch doesn't work for pre-release versions.
+* `3.0.0 rollForward = Minor` would roll forward to `3.1.0` even if `3.0.1-preview` is also available on the machine (and technically is closer to the desired version).
+* `2.0.0 rollForward = LatestMajor` would roll forward to `3.0.0` even if `3.0.1-preview` is also available on the machine.
+* `3.0.0 rollForward = Minor` would roll forward to `3.0.1-preview` if that's the only version available on the machine.
+
+Pros
+* Installing pre-release version doesn't affect apps which use release version (unless it's needed to make the app work).
+* Doesn't impose any implicit expectations on the quality of pre-release versions as typically pre-release would only be used when explicitly asked for.
+* Seems to match most users' expectations.
+
+Cons
+* Testing behavior of new releases with pre-release versions is not fully possible (see below).
+* Some special cases don't work:
+One special case which would not work:  
+*Component A which asks for `2.0.0 LatestMajor` is loaded first on a machine which has `3.0.0` and also `3.1.0-preview` installed. Because it's the first in the process it will resolve the runtime according to the above rules - that is prefer release version - and thus will select `3.0.0`.  
+Later on component B is loaded which asks for `3.1.0-preview LatestMajor` (for example the one in active development). This load will fail since `3.0.0` is not enough to run this component.  
+Loading the components in reverse order (B first and then A) will work since the `3.1.0-preview` runtime will be selected.*
+
+### Proposed new "pre-release" mode
+The above behavior makes sense for most users, but it makes it hard for us to test new versions of frameworks. Let's assume .NET Core 3.0 already shipped and there are apps which target `3.0.0 rollForward = Minor` (the default). The shipped framework is version `3.0.0`. Now the next patch release is being prepared and `3.0.1.preview` is produced. With the proposed (and current) behavior, there's no good way to make the apps use the new preview for testing purposes.
+
+The proposal is to add a new environment variable `DOTNET_ROLL_FORWARD_TO_PRERELEASE`. There would only be the environment variable (no command line or `.runtimeconfig.json` property). By default when it's not set or set to anything but `1` the behavior would be as described above.
```

Comment:
* Why is `DOTNET_ROLL_FORWARD_TO_PRERELEASE` a runtime setting, rather than a build time setting? 
Is the understanding here that this is a feature that is used only in development / testing scenarios?
That is, shipping apps wouldn't want to roll-forward to pre-release versions? 

* Why is this a global setting, rather than per-framework? 
Given the scenario of supporting multiple custom frameworks (not just .net and asp.net) a developer may want o to release (maybe preview versios of ) an app that rolls-forward to pre-release versions of their own framework, but not others. 

* Why is support for pre-release different from Major and Minor version specifiers? 
My thinking is that "prerelease" is simply another part of version specification. So, why not support it similar to version numbers?  `Major` `Minor` `PreviewMajor` `PreviewMinor` etc. ? 

I understand that `DOTNET_ROLL_FORWARD_TO_PRERELEASE` is simpler than supporting runtime configuration options. But we should discuss other options in the document.

Reply:
It is very important for the test scenario that we can add prerelease consideration without otherwise changing the file and without changing built app configs.

If we were to have to pick one of PreviewMajor,PreviewMinor,etc. we'd be clobbering the other settings of the app.

Reply:
In other words, we want a setting that says "run apps as they would run if the prerelease runtimes were RTM". This does not correspond to PreviewMajor or PreviewMinor.

Reply:
@nguerrera thanks for clarifying the use-case.

Reply:
I added description to the doc which should hopefully reflect the discussion here. As for why env. variable only: Currently it's the one we need, I'm not aware of a scenario which would require either CLI or config setting. As such trying to use this in CLI or config will fail, so it leaves it possible for us to add support for these other ways in the future without introducing breaking changes.

Reply:
To be honest, I'm tempted to change the name to something like `__TEST_DOTNET_ROLL_FORWARD_TO_PRERELEASE` to make it very clear that this is something developers should not need. But it seems that we may need this to be used by external users for testing new releases of our frameworks, as such the name should not sound too scary.

Reply:
Maybe simply `__TEST_DOTNET_PRERELEASE_AS_RELEASE`